File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    private static long countLessOrEqual(long x, int n) {
        long count = 0;
        for (int i = 1; i <= n; i++) {
            count += Math.min(x / i, n);
        }

        return count;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());

        long left = 1;
        long right = k;
        long answer = 0;

        while (left <= right) {
            long mid = (left + right) / 2;
            long count = countLessOrEqual(mid, n);

            if (count < k) {
                left = mid + 1;
            } else {
                answer = mid;
                right = mid - 1;
            }
        }

        System.out.println(answer);
    }
}

File: 백준/Gold/1300. K번째 수/K번째 수.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val k = br.readLine().toInt()

    var left = 1L
    var right = k.toLong()

    while (left < right) {
        val mid = (left + right) / 2
        val count = (1..n).sumOf { minOf(mid / it, n.toLong()) }

        if (count < k) {
            left = mid + 1
        } else {
            right = mid
        }
    }

    println(left)
}